### PR TITLE
Support spawning new Dart isolates from within a Flutter app

### PR DESCRIPTION
**Pull request by jason-simmons**

### CODE REVIEW

File: sky/engine/core/script/dart_init.cc [hunk 1]
```diff
@@ -51,6 +52,10 @@ extern const uint8_t* observatory_assets_archive;
 
 namespace blink {
 
+using mojo::asset_bundle::ZipAssetBundle;
+
+const char kSnapshotAssetKey[] = "snapshot_blob.bin";
```
Comment (abarth) [inline sky/engine/core/script/dart_init.cc]: Its too bad this string is duplicated in so many places.

File: sky/engine/tonic/dart_io.cc [hunk 2]
```diff
@@ -10,11 +10,21 @@
 
 namespace blink {
 
-void DartIO::InitForIsolate() {
+void DartIO::InitForIsolate(const char* script_uri) {
```
Comment (abarth) [inline sky/engine/tonic/dart_io.cc]: I'd probably take this argument as a `const std::string&`

File: sky/shell/ui/engine.cc [hunk 3]
```diff
@@ -28,8 +30,6 @@ namespace sky {
 namespace shell {
 namespace {
 
-const char kSnapshotKey[] = "snapshot_blob.bin";
```
Comment (abarth) [inline sky/shell/ui/engine.cc]: Oh, you removed it from here.  Great.

File: sky/engine/tonic/dart_io.cc [hunk 4]
```diff
@@ -10,11 +10,21 @@
...
       dart::bin::IONativeLookup,
       dart::bin::IONativeSymbol));
+
+  if (script_uri) {
+    Dart_Handle platform_type = Dart_GetType(io_lib, ToDart("_Platform"),
+                                             0, nullptr);
+    DART_CHECK_VALID(platform_type);
+    DART_CHECK_VALID(Dart_SetField(
+        platform_type, ToDart("_nativeScript"), ToDart(script_uri)));
+  }
```
Comment (abarth) [inline sky/engine/tonic/dart_io.cc]: IMHO, we should do this work in DartRuntimeHooks::Install, which is where we do all this similar stuff for the other built-in libraries.

**Comment by abarth** [comment]
LGTM, but I would move the platformScript initialization into dart_runtime_hooks.cc to keep it together with all the other places we poke into the VM.

You probably also need to add the name of the function you're calling to dart_vm_entry_points.txt so that it won't be dead-code eliminated in precompiled builds.

File: sky/engine/tonic/dart_io.cc (hunk 2 above)
Comment (jason-simmons) [inline sky/engine/tonic/dart_io.cc]: This parameter is optional for isolates that don't have a real URI (i.e. the service isolate)

**Comment by jason-simmons** [comment]
Moved the dart:io initialization into DartRuntimeHooks

File: sky/engine/tonic/dart_io.cc [hunk 5]
```diff
@@ -11,8 +11,10 @@
 namespace blink {
 
 void DartIO::InitForIsolate() {
+  Dart_Handle io_lib = Dart_LookupLibrary(ToDart("dart:io"));
+
   DART_CHECK_VALID(Dart_SetNativeResolver(
-      Dart_LookupLibrary(ToDart("dart:io")),
+      io_lib,
```
Comment (abarth) [inline sky/engine/tonic/dart_io.cc]: This diff probably isn't needed now.

File: sky/engine/bindings/dart_runtime_hooks.h [hunk 6]
```diff
@@ -18,7 +18,7 @@ class DartRuntimeHooks {
     DartIOIsolate,
   };
 
-  static void Install(IsolateType isolate_type);
+  static void Install(IsolateType isolate_type, const char* script_uri);
```
Comment (abarth) [inline sky/engine/bindings/dart_runtime_hooks.h]: You can pass an empty std::string to signal that there is no script uri.

**Comment by abarth** [comment]
LGTM

**Pull request merged**